### PR TITLE
speed up `collector()` helper

### PR DESCRIPTION
**Pull request by simonpcouch**
This helper takes up most all of the time in `collect_predictions()`. tune will see some speedup here, but `stacks::add_candidates()` especially benefits.

With `main` dev:

``` r
library(tidymodels)

res <- 
  fit_resamples(
    linear_reg(), 
    mpg ~ ., 
    bootstraps(mtcars),
    control = control_grid(save_pred = TRUE)
  )

bench::mark(
  predictions = tune:::collector(res, ".predictions"),
  metrics = tune:::collector(res, ".metrics"),
  check = FALSE
)
#> # A tibble: 2 × 6
#>   expression       min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 predictions   3.12ms    3.4ms      283.   621.4KB     6.24
#> 2 metrics       3.02ms   3.35ms      298.    36.5KB     6.22
```

With this PR:

``` r
#> # A tibble: 2 × 6
#>   expression       min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 predictions    433µs    445µs     2236.    75.7KB     6.18
#> 2 metrics        328µs    337µs     2960.    12.8KB     8.26
```

### CODE REVIEW

File: R/collect.R [hunk 1]
```diff
@@ -358,11 +358,25 @@ collector <- function(x, coll_col = ".predictions") {
...
-  x <- dplyr::select(x, dplyr::starts_with("id"), !!!keep_cols)
-  x <- tidyr::unnest(x, cols = c(dplyr::all_of(coll_col)))
+
+  id_cols <- colnames(x)[grepl("id", colnames(x))]
+  keep_cols <- c(coll_col, id_cols)
+  x <- x[, keep_cols]
+
+  res <-
+    vctrs::vec_cbind(
+      vctrs::list_unchop(x[[coll_col]]),
```
Comment (simonpcouch) [inline R/collect.R:368]: As used in tune, `coll_col` is always one of `.metrics` or `.predictions`--i.e. always has length one.

File: R/collect.R [hunk 2]
```diff
@@ -358,11 +358,25 @@ collector <- function(x, coll_col = ".predictions") {
...
+      vctrs::vec_rep_each(x[, id_cols], times = vctrs::list_sizes(x[[coll_col]]))
```
Comment (simonpcouch) [inline R/collect.R:369]: Because of the previous `select()`ion, the only other columns are ids.

File: R/collect.R [hunk 3]
```diff
@@ -358,11 +358,26 @@ collector <- function(x, coll_col = ".predictions") {
...
   arrange_cols <- c(".iter", ".config")
-  arrange_cols <- arrange_cols[(arrange_cols %in% names(x))]
-  arrange(x, !!!rlang::syms(arrange_cols))
+  arrange_cols <- arrange_cols[rlang::has_name(res, arrange_cols)]
+
+  if (length(arrange_cols) == 1) {
+    return(res[vctrs::vec_order(res[[arrange_cols]]), ])
+  }
+
+  arrange(res, !!!rlang::syms(arrange_cols))
```
Comment (DavisVaughan) [inline R/collect.R:380]:
```suggestion
  res <- vctrs::vec_slice(res, vctrs::vec_order(res[arrange_cols]))
```

You can call `vec_order()` on a data frame with multiple columns, like how `arrange()` works

**Review by DavisVaughan — APPROVED** [review]

File: R/collect.R (hunk 3 above)
Comment (simonpcouch) [reply R/collect.R:380]: Ohhhh I see how this works now, thank you.

**Pull request merged**